### PR TITLE
FOUNDATIONS: Retrieve Labs Data

### PR DESCRIPTION
**Pull request by terrypalo**

### CODE REVIEW

File: DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.Exceptions.RetrieveAll.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+using Xeptions;
+using Xunit;
+
+namespace DMX.Gatekeeper.Api.Tests.Unit.Services.Foundations.Labs
+{
+    public partial class LabServiceTests
+    {
+        [Theory]
+        [MemberData(nameof(CriticalDependencyException))]
+        public async Task ShouldThrowCriticalDependencyExceptionOnRetrievalIfCriticalErrorOccursAndLogItAsync(
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.Exceptions.RetrieveAll.cs:22]: OnRetrieve or OnRetrieveAll

File: DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.Exceptions.RetrieveAll.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+            this.dmxApiBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnRetrievalIfErrorOccursAndLogItAsync()
+        {
+            // given
+            var someMessage = GetRandomString();
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.Exceptions.RetrieveAll.cs:61]: string

File: DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.Exceptions.RetrieveAll.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+{
+    public partial class LabServiceTests
+    {
+        [Theory]
+        [MemberData(nameof(CriticalDependencyException))]
+        public async Task ShouldThrowCriticalDependencyExceptionOnRetrievalIfCriticalErrorOccursAndLogItAsync(
+            Xeption criticalDependencyException)
+        {
+            // given
+            var failedExternalLabDependencyException =
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.Exceptions.RetrieveAll.cs:26]: failedLabDependencyException

File: DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                new HttpResponseUrlNotFoundException(someResponseMessage, someMessage),
+                new HttpResponseUnauthorizedException(someResponseMessage, someMessage),
+                new HttpResponseForbiddenException(someResponseMessage, someMessage),
+            };
+        }
+
+        private Expression<Func<Xeption, bool>> SameExceptionAs(Xeption expectedException)
+        {
+            return actualExpectedAssertException =>
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api.Tests.Unit/Services/Foundations/Labs/LabServiceTests.cs:53]: just actualException?

File: DMX.Gatekeeper.Api/Models/Labs/Exceptions/FailedLabDependencyException.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+using System;
+using Xeptions;
+
+namespace DMX.Gatekeeper.Api.Models.Labs.Exceptions
+{
+    public class FailedLabDependencyException : Xeption
+    {
+        public FailedLabDependencyException(Exception innerException)
+            : base(message: "Failed lab dependency error occured, contact support.",
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api/Models/Labs/Exceptions/FailedLabDependencyException.cs:13]: occurred

File: DMX.Gatekeeper.Api/Services/Foundations/Labs/LabService.Exceptions.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
+﻿using System;
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api/Services/Foundations/Labs/LabService.Exceptions.cs:1]: Copyrights header

File: DMX.Gatekeeper.Api/Services/Foundations/Labs/LabService.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        public LabService(
+            IDmxApiBroker dmxApiBroker,
+            ILoggingBroker loggingBroker)
+        {
+            this.dmxApiBroker = dmxApiBroker;
+            this.loggingBroker = loggingBroker;
+        }
+
+        public ValueTask<List<Lab>> RetrieveAllLabsAsync() =>
+             this.TryCatch(async () => await this.dmxApiBroker.GetAllLabsAsync());
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api/Services/Foundations/Labs/LabService.cs:27]: As the TryCatch is in local scope, we may skip this and align it with the same hierarchy without indendation

File: DMX.Gatekeeper.Api/Startup.cs [hunk 8]
```diff
@@ -58,5 +61,16 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             app.UseAuthorization();
             app.UseEndpoints(endpoints => endpoints.MapControllers());
         }
+
+        private static void AddBrokers(IServiceCollection services)
```
Comment (ShriHumrudha) [inline DMX.Gatekeeper.Api/Startup.cs:65]: Recommend a separate code rub for all obvious reasons :)

**Review by AndrewMcClelland — APPROVED** [review]

**Pull request merged**